[16.0][ADD] cetmix_tower_webhook: new module

GabbasovDinar opened the pull request:
This module implements incoming webhooks for Cetmix Tower. Webhooks are authorised using customisable authenticators which can be pre-configured and reused across multiple webhooks.  Webhooks and authenticators can be exported and imported using YAML format, which makes them easily sharable.

Task: 4795

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Installable "Cetmix Tower Webhook": incoming GET/POST endpoints, reusable authenticators (IP/trusted-proxy controls), execute-as-user, per-call logging, UI and menus, YAML import/export, demo data; server-side secret validation added.

- **Documentation**
  - Full README, CONFIGURE/USAGE/CONTEXT/DESCRIPTION and in-app static help with examples and return formats.

- **Settings**
  - Admin setting to configure webhook log retention (days).

- **Tests**
  - Comprehensive unit and HTTP tests for auth, execution, logging, IP/proxy handling, content types, and retention.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

GabbasovDinar commented on cetmix_tower_webhook/__init__.py:
```diff
@@ -0,0 +1,3 @@
+from . import controllers
+from . import models
+from . import wizards
```
there's no need for that

GabbasovDinar commented on cetmix_tower_webhook/pyproject.toml:
```diff
@@ -0,0 +1,3 @@
+[build-system]
+requires = ["whool"]
+build-backend = "whool.buildapi"
```
This is auto generated file, please stop comment about it

GabbasovDinar commented on cetmix_tower_webhook/controllers/__init__.py:
```diff
@@ -0,0 +1,4 @@
+# Copyright (C) 2025 Cetmix OÜ
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from . import main
```
there's no need for that

GabbasovDinar commented on cetmix_tower_webhook/__manifest__.py:
```diff
@@ -0,0 +1,25 @@
+    "author": "Cetmix",
+    "license": "AGPL-3",
+    "installable": True,
+    "depends": ["cetmix_tower_yaml"],
+    "data": [
+        "security/ir.model.access.csv",
```
why? cetmix_tower_yaml already depends on base

GabbasovDinar commented on cetmix_tower_webhook/controllers/main.py:
```diff
@@ -0,0 +1,174 @@
+        http_method = request.httprequest.method.lower()
+
+        # Step 2. Find webhook
+        webhook = (
+            request.env["cx.tower.webhook"]
+            .sudo()
+            .search(
+                [
+                    ("endpoint", "=", endpoint),
+                    ("method", "=", http_method),
+                    ("active", "=", True),
+                ],
+            )
+        )
```
this model has sql constraint for unique records by method and endpoint

GabbasovDinar commented on cetmix_tower_webhook/views/cx_tower_webhook_log_views.xml:
```diff
@@ -0,0 +1,182 @@
+                    string="HTTP 200"
+                    domain="[('http_status','=',200)]"
+                />
+                <filter
+                    name="last_24h"
+                    string="Last 24 Hours"
+                    domain="[('create_date','&gt;=', (context_today() + ' 00:00:00'))]"
+                />
```
this filter is not needed any more

GabbasovDinar commented on cetmix_tower_webhook/models/cx_tower_webhook_eval_mixin.py:
```diff
@@ -0,0 +1,130 @@
+            raise ValidationError(_("Webhook code is empty"))
+
+        # prepare the code for evaluation
+        code_and_secrets = self.env["cx.tower.key"]._parse_code_and_return_key_values(
+            code, pythonic_mode=True, **kwargs.get("key", {})
+        )
+        secrets = code_and_secrets.get("key_values")
+        webhook_code = code_and_secrets["code"]
+
+        code = self.env["cx.tower.key"]._parse_code(
+            webhook_code, pythonic_mode=True, **kwargs.get("key", {})
+        )
```
are you sure? this is different methods

GabbasovDinar commented on cetmix_tower_webhook/tests/test_cx_tower_webhook_authenticator.py:
```diff
@@ -0,0 +1,157 @@
+        """
+        auth_token_header = "X-Token"
+        auth_token = "secret123"
+        code = f"result = {{'allowed': headers.get('{auth_token_header}') == '{auth_token}'}}"  # noqa: E501
```
it's ok

GabbasovDinar commented on cetmix_tower_webhook/models/cx_tower_webhook_eval_mixin.py:
```diff
@@ -0,0 +1,130 @@
+        if secrets and result and isinstance(result.get("message"), str):
+            result["message"] = self.env["cx.tower.key"]._replace_with_spoiler(
+                result["message"], secrets
+            )
```
done

GabbasovDinar commented on cetmix_tower_webhook/tests/test_cx_tower_webhook_authenticator.py:
```diff
@@ -0,0 +1,157 @@
+        """
+        auth_token_header = "X-Token"
+        auth_token = "secret123"
+        code = f"result = {{'allowed': headers.get('{auth_token_header}') == '{auth_token}'}}"  # noqa: E501
```
it's ok

GabbasovDinar commented on cetmix_tower_webhook/views/cx_tower_webhook_authenticator_views.xml:
```diff
@@ -0,0 +1,112 @@
+                        </page>
+                        <page name="yaml" string="YAML">
+                            <div groups="!cetmix_tower_yaml.group_export">
+                                <h3
+                                >You must be a member of the "YAML/Export" group to export data as YAML.</h3>
+                            </div>
```
no, it's ok, we should have possibility to translate

GabbasovDinar commented on cetmix_tower_webhook/static/description/index.html:
```diff
@@ -0,0 +1,515 @@
+<div class="document" id="cetmix-tower-webhook">
+<h1 class="title">Cetmix Tower Webhook</h1>
+
+<!-- !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+!! This file is generated by oca-gen-addon-readme !!
+!! changes will be overwritten.                   !!
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+!! source digest: sha256:4cbbeae15b8f14169c65c58ddcdbc90df4d974c7eb3f9ceedf7ebbcf47f5bd54
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! -->
```
no need changes it is auto generated file

GabbasovDinar commented on cetmix_tower_webhook/models/constants.py:
```diff
@@ -0,0 +1,104 @@
+
+# Default Python code used in Webhook
+DEFAULT_WEBHOOK_CODE = _(
+    """
+# You can return command result in the 'result' variable which is a dictionary:
+#   result = {"exit_code": <int, default=0>, "message": <string, default=None}
+#
+# Available variables - You can use all Cetmix Tower Python command variables except for the 'server' plus the following webhook specific ones:
+#  - payload: dictionary that contains the json payload or the GET parameters of the request
+"""
+)
```
this example from task description

GabbasovDinar commented on cetmix_tower_webhook/models/cx_tower_webhook_eval_mixin.py:
```diff
@@ -0,0 +1,132 @@
+        required=True,
+    )
+
+    def _compute_code_help(self):
+        """
+        Compute code help
+        """
+        self.code_help = self._default_eval_code_help()
```
it's work. multi record updating

GabbasovDinar commented on cetmix_tower_webhook/controllers/main.py:
```diff
@@ -0,0 +1,174 @@
+            code_status="skipped",
+        )
+
+        if not webhook:
+            log_values.update(
+                {
+                    "authentication_status": "failed",
+                    "http_status": 404,
+                }
+            )
+            return self._finalize_webhook_response(
+                message="Webhook not found",
+                error_message="Webhook not found",
+                **log_values,
+            )
+
```
fixed

GabbasovDinar commented on cetmix_tower_webhook/views/cx_tower_webhook_views.xml:
```diff
@@ -0,0 +1,164 @@
+                            <field name="active" />
+                            <field name="authenticator_id" />
+                        </group>
+                        <group>
+                            <field name="endpoint" />
+                            <field
+                                name="full_url"
+                                attrs="{'invisible': [('endpoint', '=', False)]}"
+                                widget="url"
+                            />
+                            <field name="method" />
```
PLEASE RECHECK BEFORE SENT COMMENT! It's computed not storable field, this field always readonly!!!!

GabbasovDinar commented on cetmix_tower_webhook/demo/demo_data.xml:
```diff
@@ -0,0 +1,37 @@
+    </record>
+
+    <!-- Webhooks -->
+    <record id="webhook_1" model="cx.tower.webhook">
+        <field name="name">Demo Webhook 1</field>
+        <field name="reference">demo_webhook_1</field>
+        <field name="authenticator_id" ref="webhook_authenticator_1" />
+        <field name="endpoint">demo_webhook_1</field>
+        <field name="code">result = {"exit_code": 0, "message": "OK"}</field>
+    </record>
```
PLEASE CHECK AGAIN, the method is 'post' by default

Aldeigja reviewed (approved):
Functional LGTM

ivs-cetmix commented on cetmix_tower_server/models/cetmix_tower.py:
```diff
@@ -256,3 +256,30 @@ def server_check_ssh_connection(
+
+        Args:
+            secret_value (Char): Value to validate
+            secret_reference (Char): Reference code of the secret
```
We need to ensure that this works in both cases:

- Complete inline reference is provided. Eg `#!cxtower.secret.aws_access_key!#`
- Just a reference is provided: `aws_access_key`

ivs-cetmix commented on cetmix_tower_webhook/controllers/main.py:
```diff
@@ -0,0 +1,174 @@
+                data = request.httprequest.data
+                return json.loads(data or "{}") if data else {}
+            elif http_method.upper() == "GET":
+                return dict(request.httprequest.args)
```
There may be a case when we have a POST request with JSON payload.

<img width="800" height="308" alt="image" src="https://github.com/user-attachments/assets/65cda5fd-2ef7-4dab-9ca9-e713353af96e" />

So we should add a setting to the webhook model where a user can explicitly select the payload type:

- JSON
- Form URLEncoded

```python

content_type = fileds.Selection(.....)

```

ivs-cetmix commented on cetmix_tower_webhook/controllers/main.py:
```diff
@@ -0,0 +1,217 @@
+                    request.env["ir.config_parameter"]
+                    .sudo()
+                    .get_param(
+                        "cetmix_tower_webhook.default_webhook_content_type", "json"
```
This parameter is not documented. We need either to remove it, otherwise document and make accessible via general settings.

GabbasovDinar replied:
done

ivs-cetmix commented:
/ocabot rebase

CetmixGitDrone commented:
Congratulations, PR rebased to [16.0](https://github.com/cetmix/cetmix-tower/commits/16.0).

ivs-cetmix commented on cetmix_tower_webhook/controllers/main.py:
```diff
@@ -0,0 +1,217 @@
+
+    def _get_payload_by_content_type(self, content_type):
+        """
+        Returns the payload based on the content type for POST requests.
```
Please ensure that the docsting is complete. Here and everywhere.

ivs-cetmix commented on cetmix_tower_webhook/controllers/main.py:
```diff
@@ -0,0 +1,217 @@
+
+    def _extract_webhook_request_payload(self, webhook):
+        """
+        Extracts the payload from the webhook request.
```
Please ensure that the docsting is complete. Here and everywhere.

ivs-cetmix commented on cetmix_tower_webhook/controllers/main.py:
```diff
@@ -0,0 +1,217 @@
+    )
+    def cetmix_webhook(self, endpoint, **kwargs):
+        """
+        Main entry point for webhook requests.
```
Please ensure that the docsting is complete. Here and everywhere.

ivs-cetmix commented on cetmix_tower_webhook/controllers/main.py:
```diff
@@ -0,0 +1,217 @@
+
+    def _finalize_webhook_response(self, message, **kwargs):
+        """
+        Creates log entry and returns HTTP response.
```
Please ensure that the docsting is complete. Here and everywhere.

ivs-cetmix commented on cetmix_tower_webhook/models/cx_tower_webhook_eval_mixin.py:
```diff
@@ -0,0 +1,168 @@
+
+    def _get_python_eval_odoo_objects(self, **kwargs):
+        """
+        Returns the Odoo objects of command available for the Python eval.
```
What's in the kwargs?

ivs-cetmix commented on cetmix_tower_webhook/tests/common.py:
```diff
@@ -0,0 +1,38 @@
+        super().setUp()
+
+        # Set base url for correct link generation
+        self.web_base_url = "https://test.odoo.com"
```
Let's stick to using "example.com" for domains here and everywhere.

ivs-cetmix replied:
@GabbasovDinar this is still pending

GabbasovDinar commented on cetmix_tower_webhook/controllers/main.py:
```diff
@@ -0,0 +1,217 @@
+        """
+        if content_type == "form":
+            return dict(request.httprequest.form)
+        data = request.httprequest.data
+        return json.loads(data or "{}") if data else {}
+
+    def _extract_webhook_request_headers(self):
+        """
+        Returns request headers as a dictionary.
+        """
+        return dict(request.httprequest.headers)
+
```
no need

GabbasovDinar commented on cetmix_tower_webhook/views/cx_tower_webhook_log_views.xml:
```diff
@@ -0,0 +1,177 @@
+                            <field name="create_date" />
+                        </group>
+                    </group>
+                    <notebook>
+                        <page name="request" string="Request Payload">
+                            <field
+                                name="request_payload"
+                                widget="ace"
+                                options="{'mode': 'json'}"
+                            />
+                        </page>
```
fixed

GabbasovDinar commented on cetmix_tower_webhook/tests/test_webhook_controller.py:
```diff
@@ -0,0 +1,437 @@
+        """
+        Test POST request with content_type form.
+        """
+        self.webhook_post.content_type = "form"
+        self.webhook_post.code = "result = {'exit_code': 0, 'message': 'POST FORM ok'}"
+
+        data = {"form_field": "ok"}
+        response = self.url_open(
+            self.url_for(self.webhook_post.endpoint),
+            data=data,
+            headers={"Content-Type": "application/x-www-form-urlencoded"},
+        )
```
it's works as well

GabbasovDinar commented on cetmix_tower_webhook/tests/test_webhook_controller.py:
```diff
@@ -0,0 +1,437 @@
+        """
+        Test POST request with content_type json.
+        """
+        self.webhook_post.content_type = "json"
+        self.webhook_post.code = "result = {'exit_code': 0, 'message': 'POST JSON ok'}"
+
+        data = json.dumps({"json_test": "ok"})
+        response = self.url_open(
+            self.url_for(self.webhook_post.endpoint),
+            data=data,
+            headers={"Content-Type": "application/json"},
+        )
```
it's works as well

GabbasovDinar commented on cetmix_tower_webhook/readme/CONFIGURE.md:
````diff
@@ -0,0 +1,52 @@
+result = {"exit_code": <int, default=0>, "message": <string, default=None}
+```
+
+**To configure the time for which the webhook calls logs are stored:**
````
fixed

GabbasovDinar commented on cetmix_tower_webhook/readme/CONFIGURE.md:
```diff
@@ -0,0 +1,52 @@
+- Authenticator. Select an Authenticator used for this web hook
+- Endpoint. Webhook andpoint. The complete webhook URL will be <your_tower_url>/cetmix_tower_webhooks/<endpoint>​
+- Run as User. Select a user to run the webhook from behalf of. CAREFUL! You must realise and understand what you are doing including all the possible consequences when selecting a specific user.
+- Code. Code that is used to authenticate the request. You can use all Cetmix Tower - Python command variables except for the server​ plus the following webhook specific ones:
+  - payload: dictionary that contains the json payload or the GET parameters of the request
+
```
done

GabbasovDinar commented on cetmix_tower_webhook/readme/CONFIGURE.md:
```diff
@@ -0,0 +1,52 @@
+- Code. Code that is used to authenticate the request. You can use all Cetmix Tower - Python command variables except for the server​ plus the following webhook specific ones:
+- headers: dictionary that contain the request headers
+- raw_data: string with the request raw data
+- payload: dictionary that contains the json payload or the GET parameters of the request
```
done

GabbasovDinar commented on cetmix_tower_webhook/models/cx_tower_webhook.py:
```diff
@@ -0,0 +1,181 @@
+        required=True,
+        help="How the payload is expected to be sent to this webhook: "
+        "as JSON body or as URL-encoded form data.",
+    )
+    user_id = fields.Many2one(
+        comodel_name="res.users",
+        string="Run as User",
+        help="Select a user to run the webhook from behalf of. If not set, "
+        "the webhook will run as the current user.\n"
+        "CAREFUL! You must realise and understand what you are doing including "
+        "all the possible consequences when selecting a specific user.",
+        default=SUPERUSER_ID,
+        required=True,
+    )
```
no. we need use SUPERUSER_ID

GabbasovDinar commented on cetmix_tower_server/models/cetmix_tower.py:
```diff
@@ -256,3 +256,36 @@ def server_check_ssh_connection(
+        if server_reference:
+            server = server.get_by_reference(server_reference)
+
+        # Try to extract reference from inline format using _extract_key_parts
+        key_parts = self.env["cx.tower.key"]._extract_key_parts(secret_reference)
+        if key_parts:
+            # _extract_key_parts returns a tuple: (key_type, reference).
+            # We only need the reference part here.
+            secret_reference = key_parts[1]
+
```
this issue is not related with current PR, ignore it

GabbasovDinar commented on cetmix_tower_webhook/models/constants.py:
```diff
@@ -0,0 +1,79 @@
+
+# Default Python code help used in Webhook Authenticator
+DEFAULT_WEBHOOK_AUTHENTICATOR_CODE_HELP = _(
+    """
+<h3>Help for Webhook Authenticator Python Code</h3>
+<div style="margin-bottom: 10px;">
+    <p>
+        The Python code for the webhook authenticator must return the <code>result</code> variable, which is a dictionary.<br>
+        <strong>Allowed keys:</strong>
+        <ul>
+            <li><code>allowed</code> (<b>bool</b>, required): Authentication result. <code>True</code> if allowed, <code>False</code> otherwise.</li>
+            <li><code>http_code</code> (<b>int</b>, optional): HTTP status code to return if authentication fails (default is 403).</li>
+            <li><code>message</code> (<b>str</b>, optional): Error message to show to the client.</li>
+        </ul>
+        <strong>Examples:</strong>
+        <pre style='background:#f7f7f7; padding:6px; border-radius:4px'>
+# Allow all requests
+result = {"allowed": True}
+
+# Deny with custom code and message
+result = {"allowed": False, "http_code": 401, "message": "Unauthorized request"}
+        </pre>
+    </p>
+    <strong>Available variables:</strong>
+</div>
+"""
+)
```
it will generated automatically

GabbasovDinar commented on cetmix_tower_webhook/models/constants.py:
```diff
@@ -0,0 +1,79 @@
+)
+
+# Default Python code help used in Webhook
+DEFAULT_WEBHOOK_CODE_HELP = _(
+    """
+<h3>Help for Webhook Python Code</h3>
+<div style="margin-bottom: 10px;">
+    <p>
+        The webhook Python code must set the <code>result</code> variable, which is a dictionary.<br>
+        <strong>Allowed keys:</strong>
+        <ul>
+            <li><code>exit_code</code> (<b>int</b>, optional, default=0): Exit code (0 means success, other values indicate failure).</li>
+            <li><code>message</code> (<b>str</b>, optional): Message to return in the HTTP response and log.</li>
+        </ul>
+        <strong>Example:</strong>
+        <pre style='background:#f7f7f7; padding:6px; border-radius:4px'>
+# Simple successful result
+result = {"exit_code": 0, "message": "Webhook processed successfully"}
+
+# Failure example
+result = {"exit_code": 1, "message": "Something went wrong"}
+        </pre>
+    </p>
+    <strong>Available variables:</strong>
+</div>
+"""
+)
```
it will generate automatically

GabbasovDinar commented on cetmix_tower_webhook/models/cx_tower_webhook_authenticator.py:
```diff
@@ -0,0 +1,164 @@
+                    "Dictionary containing the request payload "
+                    "(JSON for POST, params for GET)."
+                ),
+            },
+        }
+        res.update(super()._get_python_eval_odoo_objects(**kwargs))
+        return res
+
+    def _get_fields_for_yaml(self):
+        """Override to add fields to YAML export."""
+        res = super()._get_fields_for_yaml()
+        res += ["name", "code"]
+        return res
```
done

GabbasovDinar commented on cetmix_tower_webhook/models/cx_tower_webhook_authenticator.py:
```diff
@@ -0,0 +1,164 @@
+        help="Comma-separated list of IP addresses and/or subnets "
+        "(e.g. 192.168.1.10,192.168.2.0/24,10.0.0.1,2001:db8::/32,2a00:1450:4001:824::200e). "  # noqa: E501
+        "Requests from other addresses will be denied.",
+    )
```
done

GabbasovDinar commented on cetmix_tower_webhook/models/cx_tower_webhook_authenticator.py:
```diff
@@ -0,0 +1,164 @@
+                    return True
+            except ValueError:
+                try:
+                    if ip == ipaddress.ip_address(ip_entry):
```
done

GabbasovDinar commented on cetmix_tower_webhook/models/cx_tower_webhook.py:
```diff
@@ -0,0 +1,181 @@
+    @api.depends("endpoint")
+    def _compute_full_url(self):
+        """Compute full URL."""
+        base_url = self.env["ir.config_parameter"].sudo().get_param("web.base.url")
+        for rec in self:
+            rec.full_url = f"{base_url}/cetmix_tower_webhooks/{rec.endpoint}"
```
done

GabbasovDinar commented on cetmix_tower_webhook/models/cx_tower_webhook.py:
```diff
@@ -0,0 +1,181 @@
+        }
+        res.update(super()._get_python_eval_odoo_objects(**kwargs))
+        return res
+
+    def _get_fields_for_yaml(self):
+        """Override to add fields to YAML export."""
+        res = super()._get_fields_for_yaml()
+        res += [
+            "active",
+            "authenticator_id",
+            "endpoint",
+            "method",
+            "code",
+        ]
+        return res
```
done

GabbasovDinar commented on cetmix_tower_webhook/models/cx_tower_webhook_authenticator.py:
```diff
@@ -0,0 +1,196 @@
+                "import": kwargs.get("headers"),
+                "help": _("Dictionary of request headers."),
+            },
+            "raw_data": {
+                "import": kwargs.get("raw_data"),
+                "help": _("Raw body of the request (string)."),
+            },
```
fixed

ivs-cetmix commented on cetmix_tower_webhook/models/cx_tower_webhook.py:
```diff
@@ -0,0 +1,193 @@
+    )
+    endpoint = fields.Char(
+        required=True,
+        help="Webhook endpoint. The complete URL will be "
```
copy=False

GabbasovDinar replied:
done

GabbasovDinar replied:
done

ivs-cetmix commented on cetmix_tower_webhook/models/cx_tower_webhook.py:
```diff
@@ -0,0 +1,193 @@
+    )
+    user_id = fields.Many2one(
+        comodel_name="res.users",
+        string="Run as User",
```
copy=False

GabbasovDinar replied:
done

ivs-cetmix commented on cetmix_tower_webhook/models/cx_tower_webhook.py:
```diff
@@ -0,0 +1,193 @@
+        res = {
+            "headers": {
+                "import": kwargs.get("headers"),
+                "help": _("Dictionary of request headers."),
```
`Dictionary of request headers` (no trailing dot)

ivs-cetmix commented on cetmix_tower_webhook/models/cx_tower_webhook.py:
```diff
@@ -0,0 +1,193 @@
+                "import": kwargs.get("payload"),
+                "help": _(
+                    "Dictionary containing the request payload "
+                    "(JSON for POST, params for GET)."
```
Please remove the no trailing dot (`.`)

ivs-cetmix commented on cetmix_tower_webhook/models/cx_tower_webhook_authenticator.py:
```diff
@@ -0,0 +1,196 @@
+        res = {
+            "headers": {
+                "import": kwargs.get("headers"),
+                "help": _("Dictionary of request headers."),
```
Please remove the no trailing dot (`.`)

ivs-cetmix commented on cetmix_tower_webhook/models/cx_tower_webhook_authenticator.py:
```diff
@@ -0,0 +1,196 @@
+            },
+            "raw_data": {
+                "import": kwargs.get("raw_data"),
+                "help": _("Raw body of the request (bytes)."),
```
Please remove the no trailing dot (`.`)

ivs-cetmix commented on cetmix_tower_webhook/models/cx_tower_webhook_authenticator.py:
```diff
@@ -0,0 +1,196 @@
+                "import": kwargs.get("payload"),
+                "help": _(
+                    "Dictionary containing the request payload "
+                    "(JSON for POST, params for GET)."
```
Please remove the no trailing dot (`.`)

ivs-cetmix commented on cetmix_tower_webhook/models/cx_tower_webhook_eval_mixin.py:
```diff
@@ -0,0 +1,184 @@
+
+    code_help = fields.Html(
+        compute="_compute_code_help",
+        default=lambda self: self._default_eval_code_help(),
```
Do we need  the `default` here? the `compute` should handle the code rendering as well. Like [here](https://github.com/cetmix/cetmix-tower/blob/7ecccc28fde7b8b0cc8aac8e1e1260f72376fc51/cetmix_tower_server/models/cx_tower_command.py#L133)

ivs-cetmix commented on cetmix_tower_webhook/views/cx_tower_webhook_authenticator_views.xml:
```diff
@@ -0,0 +1,112 @@
+                        <page name="code" string="Code">
+                            <field
+                                name="code"
+                                widget="ace"
```
Please use the [ace_tower](https://github.com/cetmix/cetmix-tower/blob/7ecccc28fde7b8b0cc8aac8e1e1260f72376fc51/cetmix_tower_server/views/cx_tower_command_view.xml#L135) widget.

ivs-cetmix commented on cetmix_tower_webhook/views/cx_tower_webhook_views.xml:
```diff
@@ -0,0 +1,160 @@
+                        <page name="code" string="Code">
+                            <field
+                                name="code"
+                                widget="ace"
```
Please use the [ace_tower](https://github.com/cetmix/cetmix-tower/blob/7ecccc28fde7b8b0cc8aac8e1e1260f72376fc51/cetmix_tower_server/views/cx_tower_command_view.xml#L135) widget.

ivs-cetmix commented on cetmix_tower_webhook/views/res_config_settings_views.xml:
```diff
@@ -0,0 +1,30 @@
+                        </div>
+                        <div class="content-group">
+                            <div class="mt16">
+                                <field name="cetmix_tower_webhook_log_duration" />
```
Please add a link to the cron job so it could be reached easily
<img width="640" height="409" alt="image" src="https://github.com/user-attachments/assets/873823bf-1599-43bb-8b04-387c489c9331" />

GabbasovDinar replied:
but it's not a cron job, I use @api.autovacuum because it is Odoo’s built-in mechanism for daily cleanup tasks. Unlike a cron job, it runs automatically once a day without extra configuration, making it ideal for housekeeping tasks like deleting old logs.

ivs-cetmix commented on cetmix_tower_webhook/views/menuitems.xml:
Please move both webhooks and authenticators under the "Automation" submenu in settings
<img width="214" height="320" alt="image" src="https://github.com/user-attachments/assets/a7735604-8395-4786-a795-a3fef6f0ed7f" />

ivs-cetmix commented on cetmix_tower_webhook/models/cx_tower_webhook_eval_mixin.py:
```diff
@@ -0,0 +1,184 @@
+            dict: The 'result' variable from context
+        """
+        if not code:
+            raise ValidationError(_("Webhook code is empty"))
```
What if there is some code but it's commented out? Technically there will be code, but in reality it's the same `code is empty`.
I think we should simply do nothing and return the default response.

ivs-cetmix commented on cetmix_tower_webhook/readme/CONFIGURE.md:
```diff
@@ -0,0 +1,53 @@
+- Endpoint. Webhook andpoint. The complete webhook URL will be <your_tower_url>/cetmix_tower_webhooks/<endpoint>​
+- Run as User. Select a user to run the webhook on behalf of. CAREFUL! You must realize and understand what you are doing, including all the possible consequences when selecting a specific user.
+- Code. Code that processes the request. You can use all Cetmix Tower Python command variables (except for the server) plus the following webhook-specific one:
+  - payload: dictionary that contains the JSON payload or the GET parameters of the request
```
`- headers: dictionary that contains the request headers`

ivs-cetmix commented on cetmix_tower_webhook/readme/CONFIGURE.md:
````diff
@@ -0,0 +1,53 @@
+```
+
+## Configure a Webhook
+
````
Please add a warning same way as it's done [here](https://github.com/cetmix/cetmix-tower/blob/16.0/cetmix_tower_aws/readme/DESCRIPTION.md)

```
You must be a member of the "Cetmix Tower/Root" group to configure webhooks.
```

ivs-cetmix commented on cetmix_tower_webhook/readme/CONFIGURE.md:
```diff
@@ -0,0 +1,53 @@
+## Configure an Authenticator
+
```
Please add a warning same way as it's done [here](https://github.com/cetmix/cetmix-tower/blob/16.0/cetmix_tower_aws/readme/DESCRIPTION.md)

```
You must be a member of the "Cetmix Tower/Root" group to configure authenticators.
```

ivs-cetmix commented on cetmix_tower_webhook/readme/CONFIGURE.md:
```diff
@@ -0,0 +1,53 @@
+## Configure an Authenticator
+
+- Go to "Cetmix Tower > Settings  > Webhook Authenticators" and click "New".
```
To align with the new menu structure:
`Go to "Cetmix Tower > Settings  > Automation > ...`

ivs-cetmix commented on cetmix_tower_webhook/readme/CONFIGURE.md:
```diff
@@ -0,0 +1,53 @@
+
+**To configure the time for which the webhook call logs are stored:**
+
+- Go to "Cetmix Tower > Settings > General Settings"
```
To align with the new menu structure:
`Go to "Cetmix Tower > Settings  > Automation > ...`

ivs-cetmix commented on cetmix_tower_webhook/readme/CONFIGURE.md:
```diff
@@ -0,0 +1,53 @@
+- Go to "Cetmix Tower > Settings > General Settings"
+- Put a number of days into the "Keep Webhook for (days)" field. Default value is 30.
+
+Please refer to the [official documentation](https://cetmix.com/tower) for detailed configuration instructions.
```
tower.cetmix.com

GabbasovDinar commented on cetmix_tower_server/models/cetmix_tower.py:
```diff
@@ -256,3 +256,36 @@ def server_check_ssh_connection(
+    def server_validate_secret(
+        self, secret_value, secret_reference, server_reference=None
+    ):
+        """
+        Validate the provided secret value against the actual secret.
+
+        Accepts either a full inline reference (e.g. #!cxtower.secret.<REFERENCE>!#)
+        or just a <REFERENCE>.
+
+        Args:
+            secret_value (Char): Value to validate
+            secret_reference (Char): Reference code or inline reference
+            server_reference (Char, optional): Reference code of the server
+        Returns:
+            Bool: True if the value matches the secret, False otherwise
+        """
```
This comment does not relate to the current task; the comment appeared after rebase.

GabbasovDinar commented on cetmix_tower_webhook/controllers/main.py:
```diff
@@ -0,0 +1,250 @@
+            content_type="application/json",
+        )
+
+    def _get_remote_addr(self):
+        """
+        Return the remote IP address of the current request.
+
+        Returns:
+            str: Remote client IP address.
+        """
+        return request.httprequest.remote_addr
```
@ivs-cetmix Should we create this feature in this task?

ivs-cetmix replied:
@GabbasovDinar looks like a reasonable suggestion! Let's do it.

GabbasovDinar replied:
@coderabbitai go, but safety must be taken into account (so that users do not deliberately specify the required header that will give them access)

GabbasovDinar commented on cetmix_tower_webhook/demo/demo_data.xml:
```diff
@@ -0,0 +1,38 @@
+        <field name="name">Demo Webhook Authenticator 1</field>
+        <field name="reference">demo_webhook_authenticator_1</field>
+        <field name="allowed_ip_addresses">192.168.1.10,192.168.2.0/24,10.0.0.1</field>
+        <field name="code">result = {"allowed": True}</field>
```
it's works as well

GabbasovDinar commented on cetmix_tower_webhook/models/cx_tower_webhook_authenticator.py:
```diff
@@ -0,0 +1,196 @@
+        """
+        Run the authenticator code and return result.
+
+        Args:
+            raise_on_error (bool): Raise ValidationError on error if True.
+            kwargs: Additional variables passed to the code context, e.g.:
+                - "headers": request headers (dict)
+                - "raw_data": request body (str)
+                - "payload": parsed request payload (dict)
+
```
fixed

GabbasovDinar commented on cetmix_tower_webhook/models/cx_tower_webhook.py:
```diff
@@ -0,0 +1,195 @@
+        required=True,
+        help="Select an Authenticator used for this webhook.",
+    )
+    endpoint = fields.Char(
+        required=True,
+        copy=False,
+        help="Webhook endpoint. The complete URL will be "
+        "<your_tower_url>/cetmix_tower_webhooks/<endpoint>",
+    )
```
fixed

GabbasovDinar commented on cetmix_tower_webhook/readme/CONFIGURE.md:
```diff
@@ -0,0 +1,58 @@
+**Complete the following fields:**
+
+- Name. Authenticator name
+- Reference. Unique reference. Leave this field blank to auto generate it
+- Code. Code that is used to authenticate the request. You can use all Cetmix Tower - Python command variables except for the server​ plus the following webhook specific ones:
```
This was already resolved before

GabbasovDinar commented on cetmix_tower_webhook/readme/CONFIGURE.md:
```diff
@@ -0,0 +1,58 @@
+- Authenticator. Select an Authenticator used for this webhook
+- Endpoint. Webhook andpoint. The complete webhook URL will be <your_tower_url>/cetmix_tower_webhooks/<endpoint>​
+- Run as User. Select a user to run the webhook on behalf of. CAREFUL! You must realize and understand what you are doing, including all the possible consequences when selecting a specific user.
+- Code. Code that processes the request. You can use all Cetmix Tower Python command variables (except for the server) plus the following webhook-specific one:
+  - headers: dictionary that contains the request headers 
+  - payload: dictionary that contains the JSON payload or the GET parameters of the request
+
```
This was already resolved before

GabbasovDinar commented on cetmix_tower_webhook/tests/test_cx_tower_webhook.py:
```diff
@@ -0,0 +1,154 @@
+        result = self.simple_webhook.execute(raise_on_error=False)
+        self.assertEqual(result["message"], test_user.login)
+
+    def test_webhook_context_isolation(self):
+        """
+        Test that only payload is available in eval context;
+        extra kwargs are not accessible
+        """
```
This was already resolved before

GabbasovDinar commented on cetmix_tower_webhook/views/cx_tower_webhook_log_views.xml:
```diff
@@ -0,0 +1,180 @@
+        <field name="model">cx.tower.webhook.log</field>
+        <field name="arch" type="xml">
+            <form>
+                <header>
```
only create="0" by access right is not posible to edit

GabbasovDinar commented on cetmix_tower_webhook/views/cx_tower_webhook_log_views.xml:
```diff
@@ -0,0 +1,180 @@
+        <field name="name">cx.tower.webhook.log.view.tree</field>
+        <field name="model">cx.tower.webhook.log</field>
+        <field name="arch" type="xml">
+            <tree
+                decoration-info="authentication_status == 'success' and code_status == 'success'"
+                decoration-danger="authentication_status == 'failed' or code_status == 'failed'"
+                decoration-warning="authentication_status == 'not_required' or code_status == 'skipped'"
+            >
+                <field name="create_date" />
+                <field name="webhook_id" />
+                <field name="endpoint" />
+                <field name="request_method" />
+                <field name="http_status" />
+                <field name="authentication_status" />
+                <field name="code_status" />
+                <field name="user_id" />
+            </tree>
+        </field>
```
no need

GabbasovDinar commented on cetmix_tower_webhook/views/cx_tower_webhook_log_views.xml:
```diff
@@ -0,0 +1,180 @@
+                    string="Code Failed"
+                    domain="[('code_status','=','failed')]"
+                />
+                <filter
+                    name="http_200"
+                    string="HTTP 200"
+                    domain="[('http_status','=',200)]"
+                />
+                <group expand="0" string="Group By">
+                    <filter
+                        name="group_by_webhook"
+                        string="Webhook"
+                        context="{'group_by': 'webhook_id'}"
+                    />
+                    <filter
+                        name="group_by_method"
+                        string="Method"
+                        context="{'group_by': 'request_method'}"
+                    />
+                    <filter
+                        name="group_by_user"
+                        string="User"
+                        context="{'group_by': 'user_id'}"
+                    />
+                    <filter
+                        name="group_by_auth_status"
+                        string="Auth Status"
+                        context="{'group_by': 'authentication_status'}"
+                    />
+                    <filter
+                        name="group_by_code_status"
+                        string="Code Status"
+                        context="{'group_by': 'code_status'}"
+                    />
+                </group>
```
done

GabbasovDinar commented on cetmix_tower_webhook/views/cx_tower_webhook_views.xml:
```diff
@@ -0,0 +1,160 @@
+                        <group>
+                            <field name="endpoint" />
+                            <field
+                                name="full_url"
+                                attrs="{'invisible': [('endpoint', '=', False)]}"
+                                widget="CopyClipboardChar"
+                                options="{'string': 'Copy'}"
+                            />
```
it's translatable label

GabbasovDinar commented on cetmix_tower_webhook/views/cx_tower_webhook_views.xml:
```diff
@@ -0,0 +1,160 @@
+                    string="Enabled"
+                    domain="[('active', '=', True)]"
+                />
+                <filter
+                    name="filter_disabled"
+                    string="Disabled"
+                    domain="[('active', '=', False)]"
+                />
+                <group expand="0" string="Group By">
+                    <filter
+                        name="group_by_method"
+                        string="Method"
+                        context="{'group_by': 'method'}"
+                    />
+                    <filter
+                        name="group_by_authenticator_id"
+                        string="Authenticator"
+                        context="{'group_by': 'authenticator_id'}"
+                    />
+                    <filter
+                        name="group_by_user_id"
+                        string="User"
+                        context="{'group_by': 'user_id'}"
+                    />
+                </group>
```
done

GabbasovDinar commented on cetmix_tower_webhook/models/cx_tower_webhook_authenticator.py:
```diff
@@ -0,0 +1,196 @@
+    )
+    allowed_ip_addresses = fields.Text(
+        string="Allowed IPs",
+        help="Comma-separated list of IP addresses and/or subnets "
+        "(e.g. 192.168.1.10,192.168.2.0/24,10.0.0.1,2001:db8::/32,2a00:1450:4001:824::200e). "  # noqa: E501
+        "Requests from other addresses will be denied.",
```
it's works as well

GabbasovDinar commented on cetmix_tower_webhook/models/cx_tower_webhook_eval_mixin.py:
```diff
@@ -0,0 +1,190 @@
+        """
+        available_libraries = self._get_python_eval_odoo_objects()
+        available_libraries.update(self._get_python_eval_libraries())
+        help_text_fragments = []
+        for key, value in available_libraries.items():
+            if key == "server":
+                # Server is not available in the webhook/authenticator eval context
+                continue
+            help_text_fragments.append(f"<li><code>{key}</code>: {value['help']}</li>")
+
+        help_text = "<ul>" + "".join(help_text_fragments) + "</ul>"
+        return f"{self._get_default_python_eval_code_help()}{help_text}"
```
no need in this case, we control all libs here

GabbasovDinar commented on cetmix_tower_webhook/models/cx_tower_webhook.py:
```diff
@@ -0,0 +1,210 @@
+                context_extra={"payload": payload, "headers": kwargs.get("headers")},
+                default_result={"exit_code": 0, "message": None},
+            )
+        except Exception as e:
+            if raise_on_error:
+                raise ValidationError(
+                    _("Webhook code execution error: %(error)s", error=e)
+                ) from e
+            result = {
+                "exit_code": 1,
+                "message": str(e),
+            }
```
no need this exception log, we  have webhook log model for this

GabbasovDinar commented on cetmix_tower_webhook/tests/test_cx_tower_webhook.py:
```diff
@@ -0,0 +1,154 @@
+        self.simple_webhook.write({"code": "result = 'not a dict'"})
+        result = self.simple_webhook.execute(raise_on_error=False)
+        self.assertEqual(result["exit_code"], 1)
+        self.assertEqual(
+            result["message"], "Webhook/Authenticator code error: result is not a dict"
+        )
```
it's works as well

GabbasovDinar commented on cetmix_tower_webhook/models/cx_tower_webhook_authenticator.py:
```diff
@@ -0,0 +1,372 @@
+        "(e.g. 192.168.1.10,192.168.2.0/24,10.0.0.1,2001:db8::/32,2a00:1450:4001:824::200e). "  # noqa: E501
+        "Requests from other addresses will be denied.",
+    )
+    trusted_proxy_ips = fields.Char(
+        string="Trusted Proxy IPs",
+        help="Comma-separated list of trusted proxy IP addresses or CIDR ranges "
+        "(e.g., 10.0.0.1,192.168.1.0/24). "
+        "Only these proxies can set X-Forwarded-For headers.",
+    )
```
fixed

GabbasovDinar commented on cetmix_tower_webhook/tests/test_webhook_controller.py:
```diff
@@ -0,0 +1,606 @@
+        """
+        Test POST request to a webhook that raises an exception in code.
+        """
+        self.webhook_post.code = "raise Exception('Some error!')"
+        response = self.url_open(
+            self.url_for(self.webhook_post.endpoint),
+            data=json.dumps({}),
+            headers={"Content-Type": "application/json"},
+        )
+        self.assertEqual(response.status_code, 500)
+        self.assertIn(b"Some error!", response.content)
+
```
it's ok for tests

GabbasovDinar commented on cetmix_tower_webhook/views/cx_tower_webhook_authenticator_views.xml:
```diff
@@ -0,0 +1,116 @@
+                            <field name="reference" />
+                        </group>
+                        <group>
+                            <field
+                                name="allowed_ip_addresses"
+                                placeholder="e.g.: 192.168.1.10, 192.168.2.0/24, 2001:db8::/32, 2a00:1450:4001:824::200e"
+                            />
+                            <field
+                                name="trusted_proxy_ips"
+                                placeholder="10.0.0.1,192.168.1.0/24"
+                            />
```
I use comma only

GabbasovDinar commented on cetmix_tower_webhook/models/cx_tower_webhook_authenticator.py:
```diff
@@ -0,0 +1,371 @@
+        "(e.g. 192.168.1.10,192.168.2.0/24,10.0.0.1,2001:db8::/32,2a00:1450:4001:824::200e). "  # noqa: E501
+        "Requests from other addresses will be denied.",
+    )
+    trusted_proxy_ips = fields.Text(
+        string="Trusted Proxy IPs",
+        help="Comma-separated list of trusted proxy IP addresses or CIDR ranges "
+        "(e.g., 10.0.0.1,192.168.1.0/24). "
+        "Only these proxies can set X-Forwarded-For headers.",
+    )
```
fixed

GabbasovDinar commented on cetmix_tower_webhook/models/cx_tower_webhook_authenticator.py:
```diff
@@ -0,0 +1,371 @@
+        Returns:
+            list[str]: List of field names.
+        """
+        res = super()._get_fields_for_yaml()
+        res += ["name", "code", "allowed_ip_addresses", "trusted_proxy_ips"]
+        return res
```
reference will get from super

GabbasovDinar commented on cetmix_tower_webhook/models/cx_tower_webhook_authenticator.py:
```diff
@@ -0,0 +1,371 @@
+                    client_ip = ipaddress.ip_address(candidate)
+                except ValueError:
+                    # Fall back to immediate peer if candidate is invalid.
+                    _logger.warning(
+                        _("Invalid IP/CIDR entry: %(spec)s", spec=candidate)
+                    )
+                    pass
+
```
we don't need translate logs, so I will remove translate method

GabbasovDinar commented on cetmix_tower_webhook/views/cx_tower_webhook_authenticator_views.xml:
```diff
@@ -0,0 +1,116 @@
+                        </page>
+                        <page name="yaml" string="YAML">
+                            <div groups="!cetmix_tower_yaml.group_export">
+                                <h3
+                                >You must be a member of the "YAML/Export" group to export data as YAML.</h3>
+                            </div>
```
fixed

GabbasovDinar commented on cetmix_tower_webhook/models/cx_tower_webhook_authenticator.py:
```diff
@@ -0,0 +1,369 @@
+                ipaddress.ip_network(s, strict=False)
+                continue
+            except ValueError:
+                _logger.warning("Invalid IP/CIDR entry: %s", s)
```
fixed

GabbasovDinar commented on cetmix_tower_webhook/views/res_config_settings_views.xml:
```diff
@@ -0,0 +1,30 @@
+            ref="cetmix_tower_server.res_config_settings_view_form"
+        />
+        <field name="arch" type="xml">
+            <xpath expr="//div[@id='cetmix_tower_settings']" position="inside">
+                <div class="col-12 col-lg-6 o_setting_box">
```
it's works by autovacuum so no need cron configuration

GabbasovDinar commented on cetmix_tower_webhook/views/res_config_settings_views.xml:
```diff
@@ -0,0 +1,30 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+    <record id="res_config_settings_view_form" model="ir.ui.view">
```
no need this changes, it's works

GabbasovDinar commented on cetmix_tower_webhook/views/cx_tower_webhook_views.xml:
```diff
@@ -0,0 +1,165 @@
+                            <field name="code_help" />
+                        </page>
+                        <page name="yaml" string="YAML">
+                            <div groups="!cetmix_tower_yaml.group_export">
+                                <h3
+                                >You must be a member of the "YAML/Export" group to export data as YAML.</h3>
+                            </div>
```
it was fixed early

GabbasovDinar commented on cetmix_tower_webhook/tests/test_cx_tower_webhook_authenticator.py:
```diff
@@ -0,0 +1,143 @@
+        self.assertEqual(result.get("http_code"), 418)
+        self.assertEqual(result.get("message"), message)
+
+    def test_authenticator_returns_non_dict(self):
+        """
+        Test that authentication fails if code returns non-dict result
+        """
+        self.simple_authenticator.write({"code": "result = 'not a dict'"})
+        result = self.simple_authenticator.authenticate(
+            headers={}, raise_on_error=False
+        )
+        self.assertFalse(result["allowed"])
+        self.assertEqual(result["http_code"], 500)
+        self.assertIn("result is not a dict", result["message"])
```
it was fixed early

GabbasovDinar commented on cetmix_tower_webhook/models/res_config_settings.py:
```diff
@@ -0,0 +1,13 @@
+    _inherit = "res.config.settings"
+
+    cetmix_tower_webhook_log_duration = fields.Integer(
+        string="Keep Webhook Logs for (days)",
+        help="Set the number of days to keep webhook logs. "
```
fixed

GabbasovDinar commented on cetmix_tower_webhook/models/cx_tower_webhook_log.py:
```diff
@@ -0,0 +1,195 @@
+            .sudo()
+            .get_param("cetmix_tower_webhook.webhook_log_duration", 30)
+        )
+        cutoff = fields.Datetime.now() - timedelta(days=days)
+        logs_to_delete = self.search([("create_date", "<", cutoff)])
+        logs_to_delete.unlink()
```
fixed

GabbasovDinar commented on cetmix_tower_webhook/models/cx_tower_webhook_log.py:
```diff
@@ -0,0 +1,195 @@
+    def _compute_display_name(self):
+        """Compute display name."""
+        for rec in self:
+            rec.display_name = (
+                f"{rec.webhook_id.display_name or ''} ({rec.endpoint}) "
+                f"[{rec.http_status or ''}]"
+            )
```
endpoint is required field

GabbasovDinar commented on cetmix_tower_webhook/models/cx_tower_webhook_eval_mixin.py:
```diff
@@ -0,0 +1,190 @@
+        """
+        Return the default code help text for webhook or authenticator.
+
+        We use default because the computation method for this field
+        would not be trigged before this record is saved. And we need
+        to show the value instantly.
```
fixed

GabbasovDinar commented on cetmix_tower_webhook/models/cx_tower_webhook_authenticator.py:
```diff
@@ -0,0 +1,372 @@
+            list[ipaddress.IPv4Network|ipaddress.IPv6Network]
+        """
+        nets = []
+        if not spec:
+            return nets
+        for part in spec.split(","):
+            s = (part or "").strip()
+            if not s:
+                continue
+            try:
+                nets.append(ipaddress.ip_network(s, strict=False))
+                continue
+            except ValueError:
+                _logger.warning(
+                    "Invalid IP/CIDR entry encountered in "
+                    "trusted_proxy_ips configuration."
+                )
+                pass
+            try:
+                ip = ipaddress.ip_address(s)
+                nets.append(
+                    ipaddress.ip_network(
+                        ip.exploded + ("/32" if ip.version == 4 else "/128")
+                    )
+                )
+            except ValueError:
+                # Ignore invalid entries silently; validation is handled by constraints.
+                continue
+        return nets
```
fixed

GabbasovDinar commented on cetmix_tower_webhook/models/cx_tower_webhook_authenticator.py:
```diff
@@ -0,0 +1,372 @@
+            if not candidate:
+                candidate = self._extract_ip_from_header(headers.get("X-Real-IP"))
+            if candidate:
+                try:
+                    client_ip = ipaddress.ip_address(candidate)
+                except ValueError:
+                    # Fall back to immediate peer if candidate is invalid.
+                    _logger.warning("Invalid IP/CIDR entry")
+                    pass
```
fixed

ivs-cetmix commented on cetmix_tower_webhook/tests/test_webhook_controller.py:
```diff
@@ -0,0 +1,606 @@
+
+@tagged("-at_install", "post_install")
+class TestCxTowerWebhookController(HttpCase):
+    def setUp(self):
```
This is Odoo 16.0, so we should use setUpClass instead because it's faster

GabbasovDinar replied:
done

GabbasovDinar commented on cetmix_tower_webhook/views/res_config_settings_views.xml:
```diff
@@ -0,0 +1,30 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+    <record id="res_config_settings_view_form" model="ir.ui.view">
+        <field name="name">res.config.settings.view.form</field>
+        <field name="model">res.config.settings</field>
+        <field
+            name="inherit_id"
+            ref="cetmix_tower_server.res_config_settings_view_form"
+        />
```
cetmix_tower_yaml already depends on cetmix_tower_server, so no need to specify depends

GabbasovDinar commented on cetmix_tower_webhook/readme/CONFIGURE.md:
```diff
@@ -0,0 +1,58 @@
+## Configure an Authenticator
```
it's ok

GabbasovDinar commented on cetmix_tower_webhook/models/res_config_settings.py:
```diff
@@ -0,0 +1,13 @@
+from odoo import fields, models
```
it's not possible, no need this checking

GabbasovDinar commented on cetmix_tower_webhook/models/cx_tower_webhook.py:
```diff
@@ -0,0 +1,210 @@
+                context_extra={"payload": payload, "headers": kwargs.get("headers")},
+                default_result={"exit_code": 0, "message": None},
+            )
+        except Exception as e:
+            if raise_on_error:
+                raise ValidationError(
+                    _("Webhook code execution error: %(error)s", error=e)
+                ) from e
+            result = {
+                "exit_code": 1,
+                "message": str(e),
+            }
```
this is autoformat code by pre-commit, no need fix

GabbasovDinar commented on cetmix_tower_webhook/models/cx_tower_webhook.py:
```diff
@@ -0,0 +1,210 @@
+        help="Webhook endpoint. The complete URL will be "
+        "<your_tower_url>/cetmix_tower_webhooks/<endpoint>",
+    )
+    full_url = fields.Char(
+        compute="_compute_full_url",
+        help="Full URL of the webhook.",
+    )
```
fixed

GabbasovDinar commented on cetmix_tower_webhook/models/cx_tower_webhook.py:
```diff
@@ -0,0 +1,210 @@
+        self_with_user = self.with_user(self.user_id)
+        payload = payload or {}
+
+        try:
+            result = self_with_user._run_webhook_eval_code(
+                self_with_user.code,
+                context_extra={"payload": payload, "headers": kwargs.get("headers")},
+                default_result={"exit_code": 0, "message": None},
+            )
```
we use special key context_extra for this

GabbasovDinar commented on cetmix_tower_webhook/models/cx_tower_webhook.py:
```diff
@@ -0,0 +1,210 @@
+        res.update(super()._get_python_eval_odoo_objects(**kwargs))
+        return res
+
+    def _get_fields_for_yaml(self):
+        """Override to add fields to YAML export."""
+        res = super()._get_fields_for_yaml()
+        res += [
+            "active",
+            "authenticator_id",
+            "endpoint",
+            "method",
+            "code",
+            "content_type",
+        ]
+        return res
```
no need export user for webhooks

GabbasovDinar commented on cetmix_tower_server/models/cetmix_tower.py:
```diff
@@ -256,3 +256,36 @@ def server_check_ssh_connection(
                     }
             time.sleep(wait_time)
+
+    @api.model
+    def server_validate_secret(
+        self, secret_value, secret_reference, server_reference=None
+    ):
+        """
+        Validate the provided secret value against the actual secret.
+
+        Accepts either a full inline reference (e.g. #!cxtower.secret.<REFERENCE>!#)
+        or just a <REFERENCE>.
+
+        Args:
+            secret_value (Char): Value to validate
+            secret_reference (Char): Reference code or inline reference
+            server_reference (Char, optional): Reference code of the server
+        Returns:
+            Bool: True if the value matches the secret, False otherwise
+        """
+        server = self.env["cx.tower.server"]
+        if server_reference:
+            server = server.get_by_reference(server_reference)
+
+        # Try to extract reference from inline format using _extract_key_parts
+        key_parts = self.env["cx.tower.key"]._extract_key_parts(secret_reference)
+        if key_parts:
+            # _extract_key_parts returns a tuple: (key_type, reference).
+            # We only need the reference part here.
+            secret_reference = key_parts[1]
+
+        value = self.env["cx.tower.key"]._resolve_key_type_secret(
+            secret_reference, server_id=server.id
+        )
+        return value == secret_value
```
it's rebased changes, no need fix it

GabbasovDinar commented on cetmix_tower_server/models/cetmix_tower.py:
```diff
@@ -256,3 +256,36 @@ def server_check_ssh_connection(
+        Returns:
+            Bool: True if the value matches the secret, False otherwise
+        """
+        server = self.env["cx.tower.server"]
+        if server_reference:
+            server = server.get_by_reference(server_reference)
+
+        # Try to extract reference from inline format using _extract_key_parts
+        key_parts = self.env["cx.tower.key"]._extract_key_parts(secret_reference)
+        if key_parts:
+            # _extract_key_parts returns a tuple: (key_type, reference).
+            # We only need the reference part here.
+            secret_reference = key_parts[1]
+
+        value = self.env["cx.tower.key"]._resolve_key_type_secret(
+            secret_reference, server_id=server.id
+        )
+        return value == secret_value
```
it's rebased changes, no need fix

ivs-cetmix reviewed (approved):
Code review LGTM

GabbasovDinar committed:
[IMP] cetmix_tower_server: Add method to validate secret value

- Add posibility validate secret value by reference and optional parameters

Task: 4795

GabbasovDinar committed:
[ADD] cx_tower_webhook: New module

- Implements incoming webhooks for Cetmix Tower.
- Webhooks are authorised using customisable authenticators which can be pre-configured and reused across multiple webhooks.
- Webhooks and authenticators can be exported and imported using YAML format, which makes them easily sharable.

Task: 4795

GabbasovDinar commented on cetmix_tower_webhook/models/res_config_settings.py:
```diff
@@ -0,0 +1,13 @@
+class ResConfigSettings(models.TransientModel):
+    _inherit = "res.config.settings"
+
+    cetmix_tower_webhook_log_duration = fields.Integer(
+        string="Keep Webhook Logs for (days)",
+        help="Set the number of days to keep webhook logs. "
+        "Old logs will be deleted automatically.",
+        default=30,
+        config_parameter="cetmix_tower_webhook.webhook_log_duration",
+    )
```
no issue here

CetmixGitDrone commented:
This PR has the `approved` label and has been created more than 5 days ago. It should therefore be ready to merge by a maintainer (or a PSC member if the concerned addon has no declared maintainer). 🤖

ivs-cetmix commented:
/ocabot merge nobump

CetmixGitDrone commented:
What a great day to merge this nice PR. Let's do it!
Prepared branch [16.0-ocabot-merge-pr-344-by-ivs-cetmix-bump-nobump](https://github.com/cetmix/cetmix-tower/commits/16.0-ocabot-merge-pr-344-by-ivs-cetmix-bump-nobump), awaiting test results.

CetmixGitDrone commented:
Congratulations, your PR was merged at 61a2d8366e2db9067a7230cfbc8c54f62b445930. Thanks a lot for contributing to cetmix. ❤️
